### PR TITLE
fix: Always bind to container selected by user

### PR DESCRIPTION
**Pull request by ad1992**
fixes https://github.com/excalidraw/excalidraw/issues/5860

- [x] Always bind to container selected by user
- [x] Update the position of bound text to be just after bound text so the z-index is correct
- [x] For text tool keep the logic as its now (Select the container with highest z-index when multiple containers intersect)
- [x] Tests

### CODE REVIEW

**Comment by dwelle** [comment]
>  Don't bind to container when using text tool as no container is selected. Checked few other tools like Miro, whimsical, Eraser work similar way.

Not so sure about this one. But we shouldn't remove the double-click-on-center-to-bind behavior, which seems to now be gone as well (non-filled element).

**Comment by ad1992** [comment]
> > Don't bind to container when using text tool as no container is selected. Checked few other tools like Miro, whimsical, Eraser work similar way.
> 
> Not so sure about this one. I'd also do that in a separate PR if anything, as it's a (breaking) feature unrelated to this fix.

I am ok with creating a separate PR but the reason I think we should remove is it will have ambiguity which I am fixing in this PR since we don't know which container will be selected hence I find it more intuitive not to bind when using text tool. If we agree I will push that change first and then merge this one. 
> 
> But we shouldn't remove the double-click-on-center-to-bind behavior, which seems to now be gone as well (non-filled element).

Good catch, fixed it and added a spec for the same

File: src/element/textWysiwyg.test.tsx [hunk 1]
```diff
@@ -837,25 +871,14 @@ describe("textWysiwyg", () => {
...
-      );
-      mouse.down();
-      expect(h.elements.length).toBe(3);
-      text = h.elements[2] as ExcalidrawTextElementWithContainer;
-      editor = document.querySelector(
-        ".excalidraw-textEditorContainer > textarea",
-      ) as HTMLTextAreaElement;
-      await new Promise((r) => setTimeout(r, 0));
-      fireEvent.change(editor, { target: { value: "Whats up?" } });
-      editor.blur();
```
Comment (dwelle) [inline src/element/textWysiwyg.test.tsx:854]: why remove this test?

File: src/scene/Scene.ts [hunk 2]
```diff
@@ -150,6 +150,19 @@ class Scene {
...
   }
+
+  insertElementAtIndex(element: ExcalidrawElement, index: number) {
+    const nextElements = [
+      ...this.elements.slice(0, index + 1),
+      element,
+      ...this.elements.slice(index + 1),
+    ];
+    this.replaceAllElements(nextElements);
+  }
```
Comment (dwelle) [inline src/scene/Scene.ts:161]: this doesn't insert at index, but _after_ index.

File: src/components/App.tsx [hunk 3]
```diff
@@ -1995,10 +1994,13 @@ class App extends React.Component<AppProps, AppState> {
             isTextElement(selectedElement) ||
             isValidTextContainer(selectedElement)
           ) {
+            const container = selectedElements[0] as ExcalidrawTextContainer;
```
Comment (dwelle) [inline src/components/App.tsx:1997]: this isn't correct. `selectedElement` can be a text element (= not a container) as per above condition.

File: src/element/textWysiwyg.test.tsx (hunk 1 above)
Comment (ad1992) [reply src/element/textWysiwyg.test.tsx:854]: This test is somewhat incorrect since if a container has bound text then clicking on the center should always edit that text instead of creating a separate text element.
I had spent some time debugging this test, I will add a separate for text tool later. I am not sure why this test was working earlier, probably coz the coords were slightly different when using text tool and since width was too small (1px) so the new text was treated outside container and now since container is being computed earlier so this test started failing.  In any case this test is actually not correct so I modified it.

**Comment by ad1992** [comment]
@dwelle this PR should be ready to merge

File: src/element/textWysiwyg.test.tsx (hunk 1 above)
Comment (dwelle) [reply src/element/textWysiwyg.test.tsx:854]: checking it out (with modified assertions), it seems not only a new text is being created, but it's also being bound to the container. It can mean the test conditions are weird / test is not working correctly, or it can mean the implem code incorrectly allows for some edge case that it shouldn't (possibly both, since I can't repro in browser). So I wouldn't totally brush this off, and rather investigate this (put logs into the `startTextEditing` handlers to see why the test is creating a new text, etc.).

File: src/scene/Scene.ts [hunk 4]
```diff
@@ -153,9 +153,9 @@ class Scene {
 
   insertElementAtIndex(element: ExcalidrawElement, index: number) {
     const nextElements = [
-      ...this.elements.slice(0, index + 1),
+      ...this.elements.slice(0, index),
       element,
-      ...this.elements.slice(index + 1),
+      ...this.elements.slice(index),
```
Comment (dwelle) [inline src/scene/Scene.ts:158]: btw, this now means that if ever call this function as you did before the change, passing a result directly from `findIndex`, it will not work. So I'd add a check for `index > -1` and probably throw (so we can discover the bug instead of getting an undefined behavior).

File: src/element/textWysiwyg.test.tsx (hunk 1 above)
Comment (dwelle) [reply src/element/textWysiwyg.test.tsx:854]: it's actually due to the initial bound text being marked as deleted (due to `0` dimensions I guess). But even fixing that still breaks it, possibly the hitbox test. So screw it. JSDOM is simply too annoying.

File: src/scene/Scene.ts (hunk 4 above)
Comment (dwelle) [reply src/scene/Scene.ts]: dpme/

**Review by dwelle — APPROVED** [review]

File: src/scene/Scene.ts [hunk 5]
```diff
@@ -150,6 +150,24 @@ class Scene {
     // (I guess?)
     this.callbacks.clear();
   }
+
+  insertElementAtIndex(element: ExcalidrawElement, index: number) {
+    if (Number.isFinite(index) && index < 0) {
```
Comment (dwelle) [inline src/scene/Scene.ts:155]:
```suggestion
    if (!Number.isFinite(index) || index < 0) {
```
